### PR TITLE
Harden parallel execution

### PR DESCRIPTION
**Pull request by codito**
* Update termination condition to use sources count
* Handle a scenario where the concurrent execution manager thread can save leading
to parallel execution hang
* Send abort in case of exception during Discovery or Execution in the runner. This will ensure a new execution manager is spawned to replace in case of abort.
* Do not count tests in case a discovery aborts for a parallel discovery scenario.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs [hunk 1]
```diff
@@ -91,15 +95,22 @@ public int StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler e
...
+                var testCaseLists = testCasesBySource.Values.ToArray();
+                this.testCaseListEnumerator = testCaseLists.GetEnumerator();
+                this.availableTestSources = testCaseLists.Length;
             }
             else
             {
                 // Set the enumerator for parallel yielding of sources
                 // Whenever a concurrent executor becomes free, it picks up the next source using this enumerator
                 this.sourceEnumerator = testRunCriteria.Sources.GetEnumerator();
+                this.availableTestSources = testRunCriteria.Sources.Count();
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs]: >availableTestSources [](start = 21, length = 20)

nit: should we rename it to TestSourceCount ?

**Comment by singhsarab** [comment]
:shipit:

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs (hunk 1 above)
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs]: I am fine with name but if you are changing here then please change in ParallelProxyDiscoveryManager.cs also

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs [hunk 2]
```diff
@@ -22,6 +24,8 @@ internal class ParallelProxyExecutionManager : ParallelOperationManager<IProxyEx
         #region TestRunSpecificData
 
         private int runCompletedClients = 0;
+        private int runStartedClients = 0;
```
Comment (Faizan2304) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyExecutionManager.cs]: Can you please explain why we need this variable? Why runCompletedClients is not fulfilling our requirement?

**Review by Faizan2304 — APPROVED** [review]
I have a very small comment, rest look very fine to me 👍 :shipit:

**Pull request merged**